src/library/scala/collection/immutable/Set.scala

```diff
@@ -96,12 +96,21 @@ object Set extends IterableFactory[Set] {
 
   def from[E](it: collection.IterableOnce[E]): Set[E] =
     it match {
-      // We want `SortedSet` (and subclasses, such as `BitSet`) to
-      // rebuild themselves to avoid element type widening issues
-      case _: SortedSet[E]         => (newBuilder[E] ++= it).result()
-      case _ if it.knownSize == 0  => empty[E]
-      case s: Set[E]               => s
-      case _                       => (newBuilder[E] ++= it).result()
+      case _ if it.knownSize == 0 => empty[E]
+      // Since IterableOnce[E] launders the variance of E,
+      // identify only our implementations which can be soundly substituted.
+      // It's not sufficient to match `SortedSet[E]` to rebuild and `Set[E]` to retain.
+      case s: HashSet[E] => s
+      case s: ListSet[E] => s
+      case s: Set1[E]    => s
+      case s: Set2[E]    => s
+      case s: Set3[E]    => s
+      case s: Set4[E]    => s
+      //case s: MapOps[E @unchecked, _, _, _]#ImmutableKeySet => s
```

Comment:
It's rocket science: https://github.com/scala/bug/issues/8039

But we can cheat because anyway it's eliminated by erasure. You can use:
```suggestion
      case s: MapOps[E, Any, Map, Map[E, Any]]#ImmutableKeySet @unchecked => s
```

Btw I found a crasher:
```scala
(MapOps[E, Any, Map, Map[E, Any]] @unchecked)#ImmutableKeySet
```

Reply:
I will give you co-authored-by credit. I'm still trying to understand your earlier comment on a different PR.

I thought of the rocket science ticket, but I did not think of `Any`thing better.

Reply:
This should fix the crash: https://github.com/scala/scala/pull/10498 🤞

Reply:
I forgot to annotate the :rocket: scientist.